[MacOS] Cannot use selfsigned certifacte when unlocking hub vault

### Summary

If Cryptomator Hub runs only in an internal company network with a self signed certificate, i cannot unlock any hub vault fo the hub instance.

### Please agree to the following

- [x] I have searched [existing issues](https://github.com/cryptomator/cryptomator/issues?q=) for duplicates
- [x] I agree to follow this project's [Code of Conduct](https://github.com/cryptomator/cryptomator/blob/develop/.github/CODE_OF_CONDUCT.md)

### What software is involved?

- Operating System: macOS 15
- Cryptomator: 1.15.2
- Hub: 1.4.0


### Volume Type

None

### Steps to Reproduce

1. Deploy hub to a server which uses a selfsigned cert
2. Create a hub vault there and import it into Cryptomator
3. Unlock the vault


### Expected Behavior

Cryptomator redirects to Cryptomator hub website to authenticate.

### Actual Behavior

The error screen is shown. I cannot unlock the vault.

### Reproducibility

Always

### Relevant Log Output

```shell
Error Code RKM7:2TU1:VBR3
org.cryptomator.cryptolib.api.MasterkeyLoadingFailedException: Failed to retrieve key
at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.keyloading.hub.HubKeyLoadingStrategy.loadKey(HubKeyLoadingStrategy.java:75)
at org.cryptomator.cryptofs@2.8.0/org.cryptomator.cryptofs.CryptoFileSystems.create(CryptoFileSystems.java:51)
at org.cryptomator.cryptofs@2.8.0/org.cryptomator.cryptofs.CryptoFileSystemProvider.newFileSystem(CryptoFileSystemProvider.java:196)
at org.cryptomator.cryptofs@2.8.0/org.cryptomator.cryptofs.CryptoFileSystemProvider.newFileSystem(CryptoFileSystemProvider.java:86)
at java.base/java.nio.file.FileSystems.newFileSystem(Unknown Source)
at java.base/java.nio.file.FileSystems.newFileSystem(Unknown Source)

at org.cryptomator.cryptofs@2.8.0/org.cryptomator.cryptofs.CryptoFileSystemProvider.newFileSystem(CryptoFileSystemProvider.java:126)

at org.cryptomator.desktop@1.15.2/org.cryptomator.common.vaults.Vault.createCryptoFileSystem(Vault.java:147)

at org.cryptomator.desktop@1.15.2/org.cryptomator.common.vaults.Vault.unlock(Vault.java:166)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.keyloading.KeyLoadingStrategy.use(KeyLoadingStrategy.java:79)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.unlock.UnlockWorkflow.call(UnlockWorkflow.java:82)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.unlock.UnlockWorkflow.call(UnlockWorkflow.java:37)

at javafx.graphics@23.0.1/javafx.concurrent.Task$TaskCallable.call(Unknown Source)

at java.base/java.util.concurrent.FutureTask.run(Unknown Source)

at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source)

at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)

at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

at java.base/java.lang.Thread.run(Unknown Source)

Caused by: java.util.concurrent.ExecutionException: javax.net.ssl.SSLHandshakeException: (certificate_unknown) PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(Unknown Source)

at java.base/java.util.concurrent.CompletableFuture.reportGet(Unknown Source)

at java.base/java.util.concurrent.CompletableFuture.get(Unknown Source)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.keyloading.hub.HubKeyLoadingStrategy.loadKey(HubKeyLoadingStrategy.java:62)

... 18 more

Caused by: javax.net.ssl.SSLHandshakeException: (certificate_unknown) PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

at java.net.http/jdk.internal.net.http.HttpClientImpl.send(Unknown Source)

at java.net.http/jdk.internal.net.http.HttpClientFacade.send(Unknown Source)

at io.github.coffeelibs.tinyoauth2client@0.8.0/io.github.coffeelibs.tinyoauth2client.AuthorizationCodeGrant$WithAuthorizationCode.getAccessToken(AuthorizationCodeGrant.java:245)

at io.github.coffeelibs.tinyoauth2client@0.8.0/io.github.coffeelibs.tinyoauth2client.AuthorizationCodeGrant.authorize(AuthorizationCodeGrant.java:165)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.keyloading.hub.AuthFlowTask.call(AuthFlowTask.java:42)

at org.cryptomator.desktop@1.15.2/org.cryptomator.ui.keyloading.hub.AuthFlowTask.call(AuthFlowTask.java:14)

at javafx.graphics@23.0.1/javafx.concurrent.Task$TaskCallable.call(Unknown Source)

at java.base/java.util.concurrent.FutureTask.run(Unknown Source)

at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)

at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
... 3 more
Caused by: javax.net.ssl.SSLHandshakeException: (certificate_unknown) PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source)
at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(Unknown Source)
at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(Unknown Source)
at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(Unknown Source)
at java.base/sun.security.ssl.SSLHandshake.consume(Unknown Source)
at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(Unknown Source)
at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(Unknown Source)
at java.base/java.security.AccessController.doPrivileged(Unknown Source)
at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(Unknown Source)
at java.base/java.util.ArrayList.forEach(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate.lambda$executeTasks$3(Unknown Source)
at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate.executeTasks(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate.doHandshake(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader.processData(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader$ReaderDownstreamPusher.run(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(Unknown Source)
at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(Unknown Source)
... 3 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at java.base/sun.security.validator.PKIXValidator.doBuild(Unknown Source)
at java.base/sun.security.validator.PKIXValidator.engineValidate(Unknown Source)
at java.base/sun.security.validator.Validator.validate(Unknown Source)
at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(Unknown Source)
at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(Unknown Source)
... 22 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(Unknown Source)
at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(Unknown Source)
at java.base/java.security.cert.CertPathBuilder.build(Unknown Source)
... 27 more
```

### Anything else?

_No response_

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -1,0 +1,175 @@
+package org.cryptomator.networking;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.KeyStoreSpi;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.Vector;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class CombinedKeyStoreSpi extends KeyStoreSpi {
+
+	private final KeyStore primary;
+	private final KeyStore fallback;
+
+	public CombinedKeyStoreSpi(KeyStore primary, KeyStore fallback) {
+		this.primary = primary;
+		this.fallback = fallback;
+	}
+
+	@Override
+	public Key engineGetKey(String alias, char[] password) throws NoSuchAlgorithmException, UnrecoverableKeyException {
+		try {
+			Key key = primary.getKey(alias, password);
+			if (key == null) {
+				key = fallback.getKey(alias, password);
+			}
+			return key;
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public Certificate[] engineGetCertificateChain(String alias) {
+		try {
+			Certificate[] chain = primary.getCertificateChain(alias);
+			if (chain == null) {
+				chain = fallback.getCertificateChain(alias);
+			}
+			return chain;
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public Certificate engineGetCertificate(String alias) {
+		try {
+			Certificate cert = primary.getCertificate(alias);
+			if (cert == null) {
+				cert = fallback.getCertificate(alias);
+			}
+			return cert;
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public Date engineGetCreationDate(String alias) {
+		try {
+			Date date = primary.getCreationDate(alias);
+			if (date == null) {
+				date = fallback.getCreationDate(alias);
+			}
+			return date;
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public void engineSetKeyEntry(String alias, Key key, char[] password, Certificate[] chain) throws KeyStoreException {
+		throw new UnsupportedOperationException("Read-only KeyStore");
+	}
+
+	@Override
+	public void engineSetKeyEntry(String alias, byte[] key, Certificate[] chain) throws KeyStoreException {
+		throw new UnsupportedOperationException("Read-only KeyStore");
+	}
+
+	@Override
+	public void engineSetCertificateEntry(String alias, Certificate cert) throws KeyStoreException {
+		throw new UnsupportedOperationException("Read-only KeyStore");
+	}
+
+	@Override
+	public void engineDeleteEntry(String alias) throws KeyStoreException {
+		throw new UnsupportedOperationException("Read-only KeyStore");
+	}
+
+	@Override
+	public Enumeration<String> engineAliases() {
+		var aliases = new Vector<String>();
+		try {
+			var it1 = primary.aliases().asIterator();
+			while (it1.hasNext()) {
+				aliases.add(it1.next());
+			}
+			var it2 = fallback.aliases().asIterator();
+			while (it2.hasNext()) {
+				var alias = it2.next();
+				if (!aliases.contains(alias)) {
+					aliases.add(alias);
+				}
+			}
+			return aliases.elements();
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public boolean engineContainsAlias(String alias) {
+		try {
+			return primary.containsAlias(alias) || fallback.containsAlias(alias);
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public int engineSize() {
+		var aliases = engineAliases();
+		if (!aliases.hasMoreElements()) {
+			return 0;
+		}
+
+		var i = new AtomicInteger(0);
+		aliases.asIterator().forEachRemaining(_ -> i.incrementAndGet());
+		return i.get();
+	}
+
+	@Override
+	public boolean engineIsKeyEntry(String alias) {
+		return false;
+	}
+
+	@Override
+	public boolean engineIsCertificateEntry(String alias) {
+		return false;
+	}
+
+	@Override
+	public String engineGetCertificateAlias(Certificate cert) {
+		try {
+			String alias = primary.getCertificateAlias(cert);
+			if (alias == null) {
+				alias = fallback.getCertificateAlias(cert);
+			}
+			return alias;
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
+	}
+
+	@Override
+	public void engineStore(OutputStream stream, char[] password) throws IOException, NoSuchAlgorithmException, CertificateException {
+		throw new UnsupportedOperationException("Read-only KeyStore");
+	}
+
+	@Override
+	public void engineLoad(InputStream stream, char[] password) throws IOException, NoSuchAlgorithmException, CertificateException {
+		// Nothing to do; the real keystores are already loaded.
+	}
+}

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -170,6 +170,7 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 
 	@Override
 	public void engineLoad(InputStream stream, char[] password) throws IOException, NoSuchAlgorithmException, CertificateException {
-		// Nothing to do; the real keystores are already loaded.
+		primary.load(stream, password);
+		fallback.load(stream, password);
 	}
 }

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -21,7 +21,21 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 	private final KeyStore primary;
 	private final KeyStore fallback;
 
-	public CombinedKeyStoreSpi(KeyStore primary, KeyStore fallback) {
+	public static CombinedKeyStoreSpi create(KeyStore primary, KeyStore fallback) {
+		checkIfLoaded(primary);
+		checkIfLoaded(fallback);
+		return new CombinedKeyStoreSpi(primary, fallback);
+	}
+
+	private static void checkIfLoaded(KeyStore s) {
+		try {
+			s.aliases();
+		} catch (KeyStoreException e) {
+			throw new IllegalArgumentException("Keystore %s is not loaded.".formatted(s.getType()));
+		}
+	}
+
+	private CombinedKeyStoreSpi(KeyStore primary, KeyStore fallback) {
 		this.primary = primary;
 		this.fallback = fallback;
 	}

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -145,10 +145,6 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 	@Override
 	public int engineSize() {
 		var aliases = engineAliases();
-		if (!aliases.hasMoreElements()) {
-			return 0;
-		}
-
 		var i = new AtomicInteger(0);
 		aliases.asIterator().forEachRemaining(_ -> i.incrementAndGet());
 		return i.get();
@@ -156,12 +152,20 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 
 	@Override
 	public boolean engineIsKeyEntry(String alias) {
-		return false;
+		try {
+			return primary.isKeyEntry(alias) || fallback.isKeyEntry(alias);
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
 	}
 
 	@Override
 	public boolean engineIsCertificateEntry(String alias) {
-		return false;
+		try {
+			return primary.isCertificateEntry(alias) || fallback.isCertificateEntry(alias);
+		} catch (KeyStoreException e) {
+			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
+		}
 	}
 
 	@Override

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -11,9 +11,10 @@ import java.security.NoSuchAlgorithmException;
 import java.security.UnrecoverableKeyException;
 import java.security.cert.Certificate;
 import java.security.cert.CertificateException;
+import java.util.Collections;
 import java.util.Date;
 import java.util.Enumeration;
-import java.util.Vector;
+import java.util.LinkedHashSet;
 import java.util.concurrent.atomic.AtomicInteger;
 
 public class CombinedKeyStoreSpi extends KeyStoreSpi {
@@ -114,20 +115,11 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 
 	@Override
 	public Enumeration<String> engineAliases() {
-		var aliases = new Vector<String>();
+		var aliases = new LinkedHashSet<String>();
 		try {
-			var it1 = primary.aliases().asIterator();
-			while (it1.hasNext()) {
-				aliases.add(it1.next());
-			}
-			var it2 = fallback.aliases().asIterator();
-			while (it2.hasNext()) {
-				var alias = it2.next();
-				if (!aliases.contains(alias)) {
-					aliases.add(alias);
-				}
-			}
-			return aliases.elements();
+			primary.aliases().asIterator().forEachRemaining(aliases::add);
+			fallback.aliases().asIterator().forEachRemaining(aliases::add);
+			return Collections.enumeration(aliases);
 		} catch (KeyStoreException e) {
 			throw new IllegalStateException("At least one keystore of [%s, %s] is not initialized.".formatted(primary.getType(), fallback.getType()), e);
 		}

--- a/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
+++ b/src/main/java/org/cryptomator/networking/CombinedKeyStoreSpi.java
@@ -170,7 +170,6 @@ public class CombinedKeyStoreSpi extends KeyStoreSpi {
 
 	@Override
 	public void engineLoad(InputStream stream, char[] password) throws IOException, NoSuchAlgorithmException, CertificateException {
-		primary.load(stream, password);
-		fallback.load(stream, password);
+		// Nothing to do; the real keystores are already loaded.
 	}
 }

--- a/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
+++ b/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
@@ -19,8 +19,12 @@ public class SSLContextWithMacKeychain extends SSLContextDifferentTrustStoreBase
 	KeyStore getTruststore() throws KeyStoreException, CertificateException, IOException, NoSuchAlgorithmException {
 		var userKeyStore = KeyStore.getInstance("KeychainStore");
 		var systemRootKeyStore = KeyStore.getInstance("KeychainStore-ROOT");
-		CombinedKeyStoreSpi spi = new CombinedKeyStoreSpi(userKeyStore, systemRootKeyStore);
-		Provider dummyProvider = new Provider("CombinedKeyStoreProvider", "1.0", "Provides a combined, read-only KeyStore") {};
-		return new KeyStore(spi, dummyProvider, "CombinedKeyStoreProvider") {};
+		try {
+			CombinedKeyStoreSpi spi = CombinedKeyStoreSpi.create(userKeyStore, systemRootKeyStore);
+			Provider dummyProvider = new Provider("CombinedKeyStoreProvider", "1.0", "Provides a combined, read-only KeyStore") {};
+			return new KeyStore(spi, dummyProvider, "CombinedKeyStoreProvider") {};
+		} catch (IllegalArgumentException e) {
+			throw new KeyStoreException(e);
+		}
 	}
 }

--- a/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
+++ b/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
@@ -6,6 +6,7 @@ import java.io.IOException;
 import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
 import java.security.cert.CertificateException;
 
 /**
@@ -16,6 +17,10 @@ public class SSLContextWithMacKeychain extends SSLContextDifferentTrustStoreBase
 
 	@Override
 	KeyStore getTruststore() throws KeyStoreException, CertificateException, IOException, NoSuchAlgorithmException {
-		return KeyStore.getInstance("KeychainStore-ROOT");
+		var userKeyStore = KeyStore.getInstance("KeychainStore");
+		var systemRootKeyStore = KeyStore.getInstance("KeychainStore-ROOT");
+		CombinedKeyStoreSpi spi = new CombinedKeyStoreSpi(userKeyStore, systemRootKeyStore);
+		Provider dummyProvider = new Provider("CombinedKeyStoreProvider", "1.0", "Provides a combined, read-only KeyStore") {};
+		return new KeyStore(spi, dummyProvider, "CombinedKeyStoreProvider") {};
 	}
 }

--- a/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
+++ b/src/main/java/org/cryptomator/networking/SSLContextWithMacKeychain.java
@@ -19,6 +19,8 @@ public class SSLContextWithMacKeychain extends SSLContextDifferentTrustStoreBase
 	KeyStore getTruststore() throws KeyStoreException, CertificateException, IOException, NoSuchAlgorithmException {
 		var userKeyStore = KeyStore.getInstance("KeychainStore");
 		var systemRootKeyStore = KeyStore.getInstance("KeychainStore-ROOT");
+		userKeyStore.load(null);
+		systemRootKeyStore.load(null);
 		try {
 			CombinedKeyStoreSpi spi = CombinedKeyStoreSpi.create(userKeyStore, systemRootKeyStore);
 			Provider dummyProvider = new Provider("CombinedKeyStoreProvider", "1.0", "Provides a combined, read-only KeyStore") {};

